src/com/haskforce/actions/HaskellHindentFormatAction.java

```diff
@@ -0,0 +1,171 @@
+import com.intellij.openapi.actionSystem.CommonDataKeys;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.command.CommandProcessor;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.ExceptionUtil;
+
+/**
+ * Action that calls hindent on the buffer it is invoked for.
+ */
+public class HaskellHindentFormatAction extends AnAction implements DumbAware {
+    private static final String NOTIFICATION_TITLE = "Reformat code with Hindent";
+    private static final Logger LOG = Logger.getInstance(HaskellHindentFormatAction.class);
+
+    /**
+     * Enable the action for Haskell files.
+     */
+    @Override
+    public void update(AnActionEvent e) {
+        PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
+        boolean isHaskell = psiFile instanceof HaskellFile;
+        e.getPresentation().setEnabled(isHaskell);
+    }
+
+    /**
+     * Main entry point. Calls hindent and detects errors.
+     */
+    @Override
+    public void actionPerformed(AnActionEvent e) {
+        final PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
+        final Project project = getEventProject(e);
+        if (project == null) return;
+        if (!(psiFile instanceof HaskellFile)) return;
+        VirtualFile virtualFile = psiFile.getVirtualFile();
+        if (virtualFile == null) return;
+
+        final String groupId = e.getPresentation().getText();
+        try {
+            GeneralCommandLine commandLine = new GeneralCommandLine();
+            final String hindentPath = ToolKey.HINDENT_KEY.getPath(project);
+            final String hindentFlags = ToolKey.HINDENT_KEY.getFlags(project);
+            if (hindentPath == null || hindentPath.isEmpty()) {
+                Notifications.Bus.notify(
+                        new Notification(groupId, NOTIFICATION_TITLE,
+                                "Hindent executable path is empty"+
+                                        "<br/><a href='configureHaskellTools'>Configure</a>",
+                                NotificationType.WARNING, new HaskellToolsNotificationListener(project)), project);
+                return;
+            }
+            commandLine.setExePath(hindentPath);
+            commandLine.getParametersList().addParametersString(hindentFlags);
+
+            final VirtualFile backingFile = psiFile.getVirtualFile();
+            if (backingFile == null) return;
+
+            ApplicationManager.getApplication().saveAll();
+
+            final String commandLineString = commandLine.getCommandLineString();
+            OSProcessHandler handler = new OSProcessHandler(commandLine.createProcess(), commandLineString);
+
+            // Pipe file into the process
```

Comment:
This is the biggest difference. Pipe the file in instead of passing the filename as an argument. Hindent supports calling it on a file, but it modifies the file in-place instead of returning it as output. I figured doing it this way would work better with IntelliJ.